krm-functions/lib/nad/v1/nad.go

```diff
@@ -220,61 +280,82 @@ func (r *NadStruct) SetConfigSpec(spec *nadv1.NetworkAttachmentDefinitionSpec) e
 
 func (r *NadStruct) SetCNIType(cniType string) error {
 	if cniType != "" {
-		nadConfigStruct := r.GetNadConfig()
-		nadConfigStruct.Plugins[0].Type = cniType
-		b, err := json.Marshal(nadConfigStruct)
+		if cniType == "ipvlan" {
+			r.CniSpecType = IpVlanType
+		} else if cniType == "macvlan" {
+			r.CniSpecType = MacVlanType
+		} else if cniType == "sriov" {
+			r.CniSpecType = SriovType
+		}
+		nadConfigStruct, err := r.GetNewNadConfig()
 		if err != nil {
 			return err
 		}
-		return r.K.SetNestedString(string(b), ConfigType...)
+		for i, plugin := range nadConfigStruct.Plugins {
+			if plugin.Type == TuningType {
+				continue
+			} else {
+				nadConfigStruct.Plugins[i].Type = cniType
+			}
+		}
+		return r.SetNadConfig(nadConfigStruct)
 	} else {
 		return fmt.Errorf("unknown cniType")
 	}
 }
 
 func (r *NadStruct) SetVlan(vlanType int) error {
 	if vlanType != 0 {
-		nadConfigStruct := r.GetNadConfig()
-		nadConfigStruct.Vlan = vlanType
-		b, err := json.Marshal(nadConfigStruct)
+		nadConfigStruct, err := r.GetNewNadConfig()
 		if err != nil {
 			return err
 		}
-		return r.K.SetNestedString(string(b), ConfigType...)
+		nadConfigStruct.Vlan = vlanType
+		return r.SetNadConfig(nadConfigStruct)
 	} else {
 		return fmt.Errorf("unknown vlanType")
 	}
 }
 
 func (r *NadStruct) SetNadMaster(nadMaster string) error {
 	if nadMaster != "" {
-		nadConfigStruct := r.GetNadConfig()
-		nadConfigStruct.Plugins[0].Master = nadMaster
-		b, err := json.Marshal(nadConfigStruct)
+		nadConfigStruct, err := r.GetNewNadConfig()
```

Comment:
Why isn't it `GetExistingNadConfig`? Intuitively it should only change the "master" field and leave everything else untouched. Isn't it the case? If it's not, then I recommend  adding a comment documenting that the setter works an unusual way.

Reply:
actually you are right, I dont know what I was thinking :D
Thanks